functions/access.js

```diff
@@ -0,0 +1,136 @@
+const { uuid } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  if (!venueId || !uid || !token) return false;
+
+  const venueRef = admin.firestore().collection("venues").doc(venueId);
+  const accessRef = admin.firestore().collection("accessgranted").doc(uid);
+
+  return await admin
+    .firestore()
+    .runTransaction(async (transaction) => {
+      const venue = await transaction.get(venueRef);
+      const granted = await transaction.get(accessRef);
+
+      if (!venue.exists) {
+        return false;
+      }
+      if (!granted.exists || !granted.tokens) {
+        return false;
+      }
+
+      if (Object.keys(granted.tokens).includes(token)) {
+        // Record that the token was checked
+        if (!granted.tokens[token].usedAt) {
+          granted.tokens[token].usedAt = [];
+        }
+        granted.tokens[token].usedAt.push(Date.now());
+        transaction.update(accessRef, granted);
+
+        return true;
+      }
+
+      return false;
+    })
+    .catch((e) => {
+      console.log("Transaction failure:", e);
+      return false;
+    });
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = await getAccessDoc(venueId, "password");
+  if (!access.exists || !access.data().password) {
+    return false;
+  }
+  return passwordsMatch(access.data().password, password);
+};
+
+const isValidEmail = async (venueId, email) => {
+  const access = await getAccessDoc(venueId, "emails");
+  if (!access.exists || !access.data().emails) {
+    return false;
+  }
+  return access.data().emails.includes(email.trim().toLowerCase());
+};
+
+const isValidCode = async (venueId, code) => {
+  const access = getAccessDoc(venueId, "code");
+  if (!access.exists || !access.data().codes) {
+    return false;
+  }
+  return access.data().codes.includes(code.trim());
+};
+
+const createToken = async (venueId, uid, password, email, code) => {
+  if (!venueId || !uid || !password || !email || !code) return {};
+
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+
+  const token = uuid();
+  const grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
+
+  const tokenData = {
+    usedAt: [Date.now()],
+    password: password || null,
+    email: email || null,
+    code: code || null,
+  };
+
+  if (grantedDoc.exists) {
+    await grantedDoc.update({ [`tokens.${token}`]: tokenData });
+  } else {
+    const granted = { tokens: { [token]: tokenData } };
+    await grantedDoc.set(granted);
+  }
```

Comment:
I haven't used this before myself, but wondering if the `{ merge: true }` option of `.set()` would be of use here, to combine them into a single thing?

- https://firebase.google.com/docs/firestore/manage-data/add-data#set_a_document
  - > If the document does not exist, it will be created. If the document does exist, its contents will be overwritten with the newly provided data, **unless you specify that the data should be merged into the existing document, as follows..**
  - > If you're not sure whether the document exists, **pass the option to merge the new data with any existing document to avoid overwriting entire documents.**